components/cloud/gcp/src/kms.rs

```diff
@@ -0,0 +1,423 @@
+    #[serde(with = "serde_base64_bytes")]
+    plaintext: Vec<u8>,
+    #[serde(deserialize_with = "deseralize_u32_from_str")]
+    plaintext_crc32c: u32,
+}
+
+fn check_crc32(data: &[u8], expected: u32) -> StdResult<(), Crc32Error> {
+    let crc = crc32c::crc32c(data);
+    if crc != expected {
+        return Err(Crc32Error { expected, got: crc });
+    }
+    Ok(())
+}
+
+pub struct Crc32Error {
+    expected: u32,
+    got: u32,
+}
+
+impl fmt::Debug for Crc32Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self)
+    }
+}
+
+impl fmt::Display for Crc32Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "crc32 mismatch, expected: {}, got: {}",
+            self.expected, self.got
+        )
+    }
+}
+
+impl std::error::Error for Crc32Error {}
+
+impl From<Crc32Error> for CloudError {
+    fn from(e: Crc32Error) -> Self {
+        Self::KmsError(KmsError::Other(Box::new(e)))
+    }
+}
+
+mod serde_base64_bytes {
+    use serde::{Deserialize, Deserializer, Serializer};
+
+    // deserialize bytes from base64 encoded string.
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<Vec<u8>, D::Error>
+    where
+        D: Deserializer<'de>,
+        D::Error: serde::de::Error,
+    {
+        let v = String::deserialize(deserializer)?;
+        base64::decode(v)
+            .map_err(|e| serde::de::Error::custom(format!("base64 decode failed: {:?}", e,)))
+    }
+
+    // serialize bytes with base64 encoding.
+    pub fn serialize<S>(data: &Vec<u8>, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let str_data = base64::encode(data);
+        serializer.serialize_str(&str_data)
+    }
+}
+
+fn deseralize_u32_from_str<'de, D>(deserializer: D) -> StdResult<u32, D::Error>
+where
+    D: Deserializer<'de>,
+    D::Error: serde::de::Error,
+{
+    let v = String::deserialize(deserializer)?;
+    v.parse().map_err(|e| {
+        serde::de::Error::custom(format!("case crc32 string '{}' as u32 failed: {:?}", &v, e,))
+    })
+}
+
+#[derive(Clone, Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct GenRandomBytesReq {
+    length_bytes: usize,
+    // TODO: Change to enum
+    protection_level: String,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct GenRandomBytesResp {
+    #[serde(with = "serde_base64_bytes")]
+    data: Vec<u8>,
+    #[serde(deserialize_with = "deseralize_u32_from_str")]
+    data_crc32c: u32,
+}
+
+#[cfg(test)]
+mod tests {
+    use cloud::kms::{Location, SubConfigGcp};
+
+    use super::*;
+
+    #[test]
+    fn test_new_gcp_kms() {
+        for bad_key in [
+            "abc",
+            "projects/test-project/locations/us-west-2/keyRings/tikv-gpc-kms-test/cryptoKeys/gl-dev-test//",
+            "projects/test-project/locations/nam8/keyRings/tikv-gpc-kms-test/cryptoKeys/gl-dev-test/cryptoKeyVersions/1",
+        ] {
```

Comment:
Is the specified configuration a self-defined and private key ? If the private resource was released, this case might fail.

Recommend using:
- public GCP encryption key
- using valid mocked function for tests, like the implementation in AWS: https://github.com/tikv/tikv/blob/d447120e32741f5de00fbb7c183c788e52871ab1/components/cloud/aws/src/kms.rs#L245-L249
- or add annotations just like https://github.com/tikv/tikv/blob/d447120e32741f5de00fbb7c183c788e52871ab1/components/cloud/azure/src/kms.rs#L303C1-L308.

Reply:
It's just a mock key the matches the google key_id pattern. Because the token provider in GcpClient is not a dynamic interface, so we can't mock it here.  In the current impl, build a client doesn't validate the `key_id` or credential, it's postponed to the first real request.

Reply:
Got. I misread it.